### DISCUSSION TITLE
How to force inclusion of some utility classes?

### DISCUSSION THREAD

**Discussion by marshallswain** [windicss/windicss#29] [Q&A] [answered]
If I have some utility classes that are only used dynamically (in this case, they're pulled from a JSON file), is there a way to configure Windi CSS to include those classes in the output?

For sure, this same code didn't work in the final output from Tailwind, since purge would wipe out the classes.  Since Windi CSS only includes what we tell it to include, it would be nice to have a simple way to specify classes to always include.

P.S., for my simple scenario, which only uses a few classes, the simple and dirty workaround was to add an element like this to make sure the classes I need are found by the compiler:

```html
<div class="hidden rotate-0 rotate-1 rotate-2 rotate-3 -rotate-1 -rotate-2 -rotate-3" />
```

Since I want to enable these programmatically, I'm picturing an option in the config like this:

```js
// Where each record contains a `classes` attribute.
const records = require('./my-data.json')

// Combine and flatten the classes
const classesFromRecords = records.map(r => r.classes).flat()

// Truncated config
{
  alwaysInclude: classesFromRecords
}
```

**Comment by antfu** [comment, marked as answer]
If you are using Vite, you can do it this way
```js
//  vite.config.js

const records = require('./my-data.json')
const classesFromRecords = records.map(r => r.classes).flat()

export default {
  plugins: [
    ...WindiCSS({ safelist: classesFromRecords }),
  ]
}
```
